Remove all references to docker, using the runc method works fine.

rhatdan opened the pull request:
This fixes the issue we saw with the /tmp being mounted outside of
the container using --tmpfs but not being seen by oci-systemd-hook

Also using this method see's the /run/secrets patch so we don't need
the special handling.

Also eliminates a whole bunch of code which is always good.

mrunalp commented:
This didn't work for me with docker. My fedora container with httpd fails to come up.

rhatdan commented:
Could you attach your Dockerfile used to build httpd?

mrunalp commented:
```
FROM fedora:24
ENV container docker
RUN dnf install -y httpd procps-ng && dnf clean all && systemctl enable httpd
STOPSIGNAL SIGRTMIN+3
EXPOSE 80
CMD [ "/sbin/init" ]
```

mrunalp commented:
FWIW, I see that the cgroup mounts are somehow missing with this PR.

rhatdan commented:
Ok so that might be something different between the configs with and without docker.  I will look into it.

rhatdan commented:
I can't get systemd working with the old oci-systemd-hook either.

mrunalp commented:
The check for contains_mount ( /sys/fs/cgroup ) is triggered now and since it is present in runc config.json that we parse, it is skipping over cgroup mount setup.

rhatdan commented:
Right, Why are we mounting that into every container?

mrunalp commented:
That is for the introspection so the containers can look at their own memory limits (for JVM tuning for e.g.).

rhatdan commented:
RIght, but it also lets them examine other processes running on the host.  I guess this is being mounted rbind, since it looks like it is all mounted readonly into the container.

mrunalp commented:
Container's own subtree is read-only bind mounted, so it can't see other processes cgroups.

mrunalp commented on src/systemdhook.c:
```diff
@@ -570,26 +478,39 @@ static int prestart(const char *rootfs,
-		 * initialize libmount
-		 */
-		mnt_init_debug(0);
+		if ((umount(dest) == -1) && errno != EINVAL) {
```
We should umount /sys/fs/cgroup/systemd only.

mrunalp commented on src/systemdhook.c:
```diff
@@ -570,26 +478,39 @@ static int prestart(const char *rootfs,
 
-		tb = parse_tabfile("/proc/self/mountinfo");
-		if (!tb) {
+		/* Mount tmpfs at new temp directory */
```
Not required since the docker cgroup mount takes care of this already. (Maybe add a check for runc?)

mrunalp commented on src/systemdhook.c:
```diff
@@ -570,26 +478,39 @@ static int prestart(const char *rootfs,
+			pr_perror("Failed to mount on %s", systemd_dest);
 			return -1;
 		}
+		if (strcmp("", mount_label)) {
```
This block should be commented out till we have kernel support.

mrunalp commented:
Applying this diff addresses my comments and works fine on testing:
```
diff --git a/src/systemdhook.c b/src/systemdhook.c
index 14a3ad5..e36a00e 100644
--- a/src/systemdhook.c
+++ b/src/systemdhook.c
@@ -479,19 +479,11 @@ static int prestart(const char *rootfs,
        }

        if (!contains_mount(config_mounts, config_mounts_len, "/sys/fs/cgroup/systemd")) {
-               _cleanup_free_ char *dest = NULL;
-               rc = asprintf(&dest, "%s/%s", rootfs, "/sys/fs/cgroup");
                _cleanup_free_ char *systemd_dest = NULL;
-               rc = asprintf(&systemd_dest, "%s/%s", dest, "systemd");
+               rc = asprintf(&systemd_dest, "%s/%s", rootfs, "/sys/fs/cgroup/systemd");

-               if ((umount(dest) == -1) && errno != EINVAL) {
-                       pr_perror("Failed to umount %s", dest);
-               }
-
-               /* Mount tmpfs at new temp directory */
-               if (mount("tmpfs", dest, "tmpfs", MS_NODEV|MS_NOSUID, options) == -1) {
-                       pr_perror("Failed to mount tmpfs at %s", tmp_dir);
-                       return -1;
+               if (umount(systemd_dest) == -1) {
+                       pr_perror("Failed to umount %s", systemd_dest);
                }

                if (makepath(systemd_dest, 0755) == -1) {
@@ -505,12 +497,15 @@ static int prestart(const char *rootfs,
                        pr_perror("Failed to mount on %s", systemd_dest);
                        return -1;
                }
+               
+               /*
                if (strcmp("", mount_label)) {
                        if (setfilecon(systemd_dest, mount_label) < 0) {
                                pr_perror("Failed to set SELinux labels on %s", systemd_dest);
                                return -1;
                        }
                }
+               */
        }

        if (!contains_mount(config_mounts, config_mounts_len, "/etc/machine-id")) {
```

mrunalp commented:
Looks fine now.

rhatdan commented:
I was talking to @rhvgoyal, and he believes the way this really should be done, is we should be bind mounting in the scope of the parent of the container on top of /sys/fs/cgroup/systemd.  But this might be something we want to put into runc.  We should also be turning on the cgroup namespace.

mrunalp commented:
> I was talking to @rhvgoyal, and he believes the way this really should be done, is we should be bind mounting in the scope of the parent of the container on top of /sys/fs/cgroup/systemd. 

Each systemd instance needs its own named cgroup mount (and nspawn also does this as far as I can tell from looking at the code):
```
systemd will also maintain its own, private, controller-less, named control group hierarchy which is mounted to /sys/fs/cgroup/systemd/. 
This hierarchy is private property of systemd, and other software should not try to interfere with it. 
This hierarchy is how systemd makes use of the naming and grouping feature of control groups (A) without actually requiring any kernel controller enabled for that.
```
Also, trying to bind mount the parent scope won't work with user namespaces when containers with different mappings have the same parent scope.

> We should also be turning on the cgroup namespace.

runc has support for cgroup namespaces. Just need to set it in config.json.

rhatdan commented:
So this is a docker issue, that it does not turn on cgroup namespace.

Not sure if you are agreeing or disagreeing.
So what I am seeing is that /sys/fs/cgroup/systemd is being mounted into the container.  And I know this is wrong.  We don't want systemd inside of the container writing to the same location as systemd outside of the container.

mrunalp commented:
> So what I am seeing is that /sys/fs/cgroup/systemd is being mounted into the container. And I know this is wrong. We don't want systemd inside of the container writing to the same location as systemd outside of the container.

Yes, that's the current behavior. However with this PR we are giving the container its own named cgroup mount so it doesn't access or use the host mount. 
Also, I will remove the bind mount to /sys/fs/cgroup/systemd from runc as it doesn't help like the other bind mounts do. (the one that we umount in this PR).

hogarthj commented:
To keep things aligned between bugzilla and here ... 

So I've built oci-systemd-hook with the current koji build and this applied on top of it:

https://patch-diff.githubusercontent.com/raw/projectatomic/oci-systemd-hook/pull/49.patch

From the three issues I raised recently:

bug 1406435 - this is still fixed with this patch applied

bug 1406830 - this bug is not fixed with this patch

docker run -d --tmpfs /tmp:rw,mode=1777,size=2G --name systemd-test systemd-test
tmpfs    7.8G     0  7.8G   0% /tmp

bug 1419040 - this bug is still fixed with this patch applied

rhatdan commented:
@mrunalp is OCI-systemd-hook running in the containers mount namespace?

mrunalp commented:
@rhatdan Yes, we join the containers mount namespace in the oci-systemd-hook.

mrunalp commented on src/systemdhook.c:
```diff
@@ -570,25 +512,32 @@ static int prestart(const char *rootfs,
-		if (rc == -1) {
-			return -1;
+		if (strcmp("", mount_label)) {
+			if (setfilecon(dest, mount_label) < 0) {
```
Still need to comment out this section.

rhatdan replied:
Is this section failing or just reporting that it can't set the labels. I had hoped the kernel would return ENOTSUP and we could keep it in place for when cgroup gains support.

mrunalp replied:
This is actually fine. I missed seeing the logic for ENOTSUP further down.

mrunalp commented:
@hogarthj The tmpfs size issue that you listed doesn't occur for me with this PR. 

```
 docker run --rm -ti --tmpfs /tmp:rw,mode=1777,size=1G fedora-httpd
```

```
[root@8ede7a4c752d /]# mount | grep /tmp  
tmpfs on /tmp type tmpfs (rw,nosuid,nodev,noexec,relatime,context="system_u:object_r:container_file_t:s0:c329,c869",size=1048576k)
```

rhatdan commented:
@mrunalp We need to implement Lennarts feedback, when I get back from PTO.

rhatdan commented:
@mrunalp I believe this implements Lennart's feedback.  PTAL.

mrunalp commented:
okay, sure.

mrunalp commented on src/systemdhook.c:
```diff
@@ -218,6 +139,32 @@ static char *get_file_contents(const char *path) {
 	return strdup(buffer);
 }
 
+DIR *dir;
```
Can we move these inside the function?

mrunalp commented:
Besides the one comment, this looks fine.

rhatdan commented:
@mrunalp Could you take another pass, I cleaned this up in a different place and thought I was pushing it here.  Now I have implemented all of Lennarts thoughts.

mrunalp commented:
okay, sure.

mrunalp commented:
Failed in my testing with docker on Fedora 25
```
[root ~]# docker run -it --rm fedora-httpd
/usr/bin/docker-current: Error response from daemon: invalid header field value "oci runtime error: container_linux.go:247: starting container process caused \"process_linux.go:334: running prestart hook 2 caused \\\"error running hook: exit status 1, stdout: , stderr: \\\"\"\n".
```

mrunalp commented:
I see this in the journal:
```
Failed to mount /sys/fs/cgroup/systemd/system.slice/docker-58db0b9d0d178f7b2c546663$37c7d5949f598958b5fe612a8dffd52d0822468.scope on /var/lib/docker/devicemapper/mnt/f33adb9413cdc90f6acfccf428db2e881876284a6715ba963a01cc8d2282a7c9/rootfs//sys/fs/cgroup/sys$emd/system.slice/docker-58db0b9d0d178f7b2c546663237c7d5949f598958b5fe612a8dffd52d0822468.scope: No such file or directory
Mar 06 12:33:34 dhcp-16-177.sjc.redhat.com oci-systemd-hook[21849]: systemdhook <error>: Failed to bind mount /sys/fs/cgroup/systemd on /var/lib/docker/devicemapper/mnt/f33$db9413cdc90f6acfccf428db2e881876284a6715ba963a01cc8d2282a7c9/rootfs//sys/fs/cgroup/systemd/system.slice/docker-58db0b9d0d178f7b2c546663237c7d5949f598958b5fe612a8dffd52d0822$68.scope: No such file or directory
```

mrunalp commented:
With latest, I see this:
```
[root@dhcp-16-101 ~]# docker run -it --rm fedora-httpd
/usr/bin/docker-current: Error response from daemon: invalid header field value "oci runtime error: container_linux.go:247: starting container process caused \"process_linux.go:334: running prestart hook 1 caused \\\"error running hook: exit status 1, stdout: , stderr: \\\"\"\n".
[root@dhcp-16-101 ~]# docker run -it --rm fedora-httpd
systemd 229 running in system mode. (+PAM +AUDIT +SELINUX +IMA -APPARMOR +SMACK +SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +SECCOMP +BLKID +ELFUTILS +KMOD +IDN)
Detected virtualization docker.
Detected architecture x86-64.

Welcome to Fedora 24 (Twenty Four)!

Set hostname to <3d6b9270976f>.
Failed to install release agent, ignoring: Permission denied
Failed to create /system.slice/docker-3d6b9270976f662cca03a480b00a35187744773ea7bed270e946bcd7ab869541.scope/init.scope control group: Permission denied
Failed to allocate manager object: Permission denied
[!!!!!!] Failed to allocate manager object, freezing.
Freezing execution.
```

rhatdan commented:
Ok that is getting closer.  I think the issue here is SELinux.  I am building a Fedora systemd container now.

rhatdan commented:
```
# getsebool -a | grep cgroup
container_manage_cgroup --> off
```
Try to enable this
```
# setsebool -P container_manage_cgroup 1
```

Which will allow systemd to write to the cgroup file system

mrunalp commented:
Ahh, okay. I will try that out.

mrunalp commented:
okay, works now. I will go through the code.

mrunalp commented on src/systemdhook.c:
```diff
@@ -273,6 +227,40 @@ static char *get_process_cgroup_subsystem_path(int pid, const char *subsystem) {
+		return -1;
+	}
+	/*
+	 * add all children to the output table
```
This comment seems out of place.

rhatdan replied:
Fixed

mrunalp commented on src/systemdhook.c:
```diff
@@ -570,28 +551,50 @@ static int prestart(const char *rootfs,
+		pr_perror("Failed to bind mount %s on %s", CGROUP_SYSTEMD, systemd_named_path);
+		return -1;
+	}
+	if (chcon(systemd_named_path,mount_label) < 0) {
```
nit: space after comma

rhatdan replied:
Fixed

mrunalp commented on src/systemdhook.c:
```diff
@@ -259,7 +206,14 @@ static char *get_process_cgroup_subsystem_path(int pid, const char *subsystem) {
-			rc = asprintf(&subsystem_path, "%s/%s%s", CGROUP_ROOT, subsystem, path);
+			const char *subpath = strchr(subsystem, '=');
+			if (subpath == NULL) {
+				subpath=subsystem;
```
nit: space around =

rhatdan replied:
Fixed

mrunalp commented on src/systemdhook.c:
```diff
@@ -259,7 +206,14 @@ static char *get_process_cgroup_subsystem_path(int pid, const char *subsystem) {
+			} else {
+				subpath++;
+			}
+			pr_pdebug("PATH: %s", subpath);
```
Should we take out the debug?

rhatdan replied:
Removed

rhatdan commented:
@mrunalp All cleaned up.

mrunalp commented on src/systemdhook.c:
```diff
@@ -273,6 +227,37 @@ static char *get_process_cgroup_subsystem_path(int pid, const char *subsystem) {
 	return NULL;
 }
 
+/* reads filesystems from @tb (libmount) looking for cgroup file systems
```
This comment looks incorrect. We are bind mounting cgroup root and ensuring that systemd_path is created before making the mount read-only.

mrunalp commented on src/systemdhook.c:
```diff
@@ -570,28 +548,50 @@ static int prestart(const char *rootfs,
+		pr_perror("Failed to create path for %s/%s", rootfs, systemd_named_path);
+		return -1;
+	}
+	if (bind_mount(named_path, systemd_named_path, false)) {
```
Should we try to make this path before we bind mount over it?

rhatdan replied:
Fixed

rhatdan replied:
Fixed

mrunalp replied:
Hmm, did you push your changes? I don't see the makepath code here.

mrunalp commented on src/systemdhook.c:
```diff
@@ -114,83 +116,28 @@ static int bind_mount(const char *src, const char *dest, int readonly) {
-			if (bind_mount(src, dest, readonly) < 0) {
-				return -1;
+			if (setfilecon (full_path, label) < 0) {
+				pr_perror("Failed to set context %s on %s", path, label);
```
nit: path and label are reversed in the error string.

rhatdan replied:
Fixed

mrunalp commented:
@rhatdan This works fine on my F25 but I see this issue on rawhide:
```
Mar 08 11:30:58 dhcp-16-129.sjc.redhat.com oci-systemd-hook[5482]: systemdhook <error>: Failed to set context /var/lib/docker/overlay2/da682f9a1aa0a1918b1b4ece8490c2863af28a95120d74de789d793c2978a2c6/merged//sys/fs/cgroup/systemd/system.slice/docker-c8d4
6a1d59763038174e78debe7c600e77a65e0c115818622500f14f0722f05f.scope on system_u:object_r:container_file_t:s0:c240,c737: Read-only file system
```

rhatdan commented:
That is weird, it should not be read-only?  It should get an error Operation Not Supported.

rhatdan commented:
Here is what I am seeing without the correct kernel.

Mar 10 07:56:26 dhcp-10-19-62-196.boston.devel.redhat.com oci-systemd-hook[30841]: systemdhook <error>: Failed to set context system_u:object_r:container_file_t:s0:c589,c758 on /var/lib/docker/overlay2/1697e42e604198644409efbb3890820634035e17ffb7a46af4a7876ae9c61393/merged//sys/fs/cgroup/systemd/system.slice/docker-43c4cfd170385d25738b923cf68bc66240c133694c71e1698989b4666da62f61.scope: Operation not supported

rhatdan commented:
I just tried it on my F25 system and I am seeing the "Operation Not supported error", I wonder if we should just not print this for now.

mrunalp commented:
@rhatdan After I updated my rawhide, I see operation not supported like you. The other changes LGTM.
